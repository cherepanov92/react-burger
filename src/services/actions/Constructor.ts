import { OrderedIngredient } from '../../utils/types';

export const ADD_INGREDIENT: 'ADD_INGREDIENT' = 'ADD_INGREDIENT';
export const REMOVE_INGREDIENT: 'REMOVE_INGREDIENT' = 'REMOVE_INGREDIENT';
export const MOVE_INGREDIENT: 'MOVE_INGREDIENT' = 'MOVE_INGREDIENT';

interface IAddIngredient {
    readonly type: typeof ADD_INGREDIENT;
    ingredient: OrderedIngredient;
}

interface IRemoveIngredient {
    readonly type: typeof REMOVE_INGREDIENT;
    ingredient: OrderedIngredient;
}

interface IMoveIngredient {
    readonly type: typeof MOVE_INGREDIENT;
    ingredient: OrderedIngredient;
    newIndex: number;
}

export type TIngredientActions = IAddIngredient | IRemoveIngredient | IMoveIngredient;