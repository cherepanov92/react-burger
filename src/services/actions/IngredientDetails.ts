import { IngredientType } from '../../utils/types';

export const ADD_INGREDIENT_DETAILS: 'ADD_INGREDIENT_DETAILS' = 'ADD_INGREDIENT_DETAILS';
export const REMOVE_INGREDIENT_DETAILS: 'REMOVE_INGREDIENT_DETAILS' = 'REMOVE_INGREDIENT_DETAILS';

interface IAddIngredientDetails {
    readonly type: typeof ADD_INGREDIENT_DETAILS;
    ingredient: IngredientType;
}

interface IRemoveIngredientDetails {
    readonly type: typeof REMOVE_INGREDIENT_DETAILS;
}

export type TIngredientDetailsActions = IAddIngredientDetails | IRemoveIngredientDetails;