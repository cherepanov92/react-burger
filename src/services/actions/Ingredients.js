import { getIngredients } from "../../utils/api";
import { APPEND_ERROR_MODAL_TYPE } from "./Modal";

export const GET_INGREDIENTS_REQUEST = 'GET_INGREDIENTS_REQUEST';
export const GET_INGREDIENTS_SUCCESS = 'GET_INGREDIENTS_SUCCESS';
export const GET_INGREDIENTS_FAILED = 'GET_INGREDIENTS_FAILED';

export const getIngredientsData = () => {
    return function(dispatch) {
        dispatch({
            type: GET_INGREDIENTS_REQUEST
        });
        getIngredients()
        .then(res => {
            if (res && res.success) {
                dispatch({
                    type: GET_INGREDIENTS_SUCCESS,
                    items: res.data
                });
            } else {
                dispatch({
                    type: GET_INGREDIENTS_FAILED
                });
            }
        })
        .catch(err => {
            dispatch({
                type: APPEND_ERROR_MODAL_TYPE,
                errorStatus: err.status
            })
        })
    };
}