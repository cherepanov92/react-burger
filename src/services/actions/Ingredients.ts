import { getIngredients } from '../../utils/api';
import { APPEND_ERROR_MODAL_TYPE } from './Modal';
import { IError, IngredientType } from '../../utils/types';
import { Dispatch } from 'react';

export const GET_INGREDIENTS_REQUEST: 'GET_INGREDIENTS_REQUEST' = 'GET_INGREDIENTS_REQUEST';
export const GET_INGREDIENTS_SUCCESS: 'GET_INGREDIENTS_SUCCESS' = 'GET_INGREDIENTS_SUCCESS';
export const GET_INGREDIENTS_FAILED: 'GET_INGREDIENTS_FAILED' = 'GET_INGREDIENTS_FAILED';

interface IGetIngredientsRequest {
    readonly type: typeof GET_INGREDIENTS_REQUEST;
}

interface IGetIngredientsSuccess {
    readonly type: typeof GET_INGREDIENTS_SUCCESS;
    items: IngredientType[];
}

interface IGetIngredientsFailed {
    readonly type: typeof GET_INGREDIENTS_FAILED;
}

export type TIngredientsActions = IGetIngredientsRequest | IGetIngredientsSuccess | IGetIngredientsFailed;

export const getIngredientsData = () => {
    return function (dispatch: Dispatch<TIngredientsActions | IError>) {
        dispatch({
            type: GET_INGREDIENTS_REQUEST
        });
        getIngredients()
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: GET_INGREDIENTS_SUCCESS,
                        items: res.data
                    });
                } else {
                    dispatch({
                        type: GET_INGREDIENTS_FAILED
                    });
                }
            })
            .catch(err => {
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                });
            });
    };
};
