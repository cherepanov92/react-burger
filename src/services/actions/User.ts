import { getUserApiData, login, logout, setUserApiData, userRegister } from '../../utils/api';
import { APPEND_ERROR_MODAL_TYPE } from './Modal';
import { IError, TUserInfoData } from '../../utils/types';
import { Dispatch } from 'react';

export const GET_USER_REQUEST: 'GET_USER_REQUEST' = 'GET_USER_REQUEST';
export const GET_USER_SUCCESS: 'GET_USER_SUCCESS' = 'GET_USER_SUCCESS';
export const GET_USER_FAILED: 'GET_USER_FAILED' = 'GET_USER_FAILED';

export const LOGOUT_USER: 'LOGOUT_USER' = 'LOGOUT_USER';

interface IGetUserRequest {
    readonly type: typeof GET_USER_REQUEST;
}

interface IGetUserSuccess {
    readonly type: typeof GET_USER_SUCCESS;
    data: TUserInfoData;
    accessToken?: string;
}

interface IGetUserFailed {
    readonly type: typeof GET_USER_FAILED;
}

interface ILogoutUser {
    readonly type: typeof LOGOUT_USER;
}

export type TUserActions = IGetUserRequest | IGetUserSuccess | IGetUserFailed | ILogoutUser;

export const loginUser = (email: string, password: string) => {
    return function (dispatch: Dispatch<IGetUserRequest | IGetUserSuccess | IGetUserFailed | IError>) {
        dispatch({
            type: GET_USER_REQUEST
        });
        login(email, password)
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: GET_USER_SUCCESS,
                        data: res.user,
                        accessToken: res.accessToken
                    });
                } else {
                    dispatch({
                        type: GET_USER_FAILED
                    });
                }
            })
            .catch(err => {
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                });
            });
    };
};

export const logoutUser = () => {
    return function (dispatch: Dispatch<ILogoutUser | IError>) {
        logout()
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: LOGOUT_USER
                    });
                }
            })
            .catch(err => {
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                });
            });
    };
};

export const getUserData = () => {
    return function (dispatch: Dispatch<IGetUserSuccess | IGetUserFailed | ILogoutUser>) {
        getUserApiData()
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: GET_USER_SUCCESS,
                        data: res.user
                    });
                } else {
                    dispatch({
                        type: GET_USER_FAILED
                    });
                }
            })
            .catch(err => {
                if (err.message === 'You should be authorised') {
                    dispatch({
                        type: LOGOUT_USER
                    });
                }
            });
    };
};

export const patchUserData = (email: string, password: string, name: string) => {
    return function (dispatch: Dispatch<IGetUserSuccess | IGetUserFailed | ILogoutUser>) {
        setUserApiData(email, password, name)
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: GET_USER_SUCCESS,
                        data: res.user
                    });
                } else {
                    dispatch({
                        type: GET_USER_FAILED
                    });
                }
            })
            .catch(err => {
                if (err.message === 'You should be authorised') {
                    dispatch({
                        type: LOGOUT_USER
                    });
                }
            });
    };
};

export const registrationUser = (email: string, password: string, name: string) => {
    return function (dispatch: Dispatch<IGetUserSuccess | IGetUserFailed | IError>) {
        userRegister(email, password, name)
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: GET_USER_SUCCESS,
                        data: res.user,
                        accessToken: res.accessToken
                    });
                } else {
                    dispatch({
                        type: GET_USER_FAILED
                    });
                }
            })
            .catch(err => {
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                });
            });
    };
};
