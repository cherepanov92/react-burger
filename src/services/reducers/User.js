export const GET_USER_REQUEST = 'GET_USER_REQUEST';
export const GET_USER_SUCCESS = 'GET_USER_SUCCESS';
export const GET_USER_FAILED = 'GET_USER_FAILED';

const UserReducer = (state = {}, action) => {
    switch (action.type) {
        case GET_USER_REQUEST: {
            return {
                ...state,
                userRequest: true,
                userFailed: false
            };
        }
        case GET_USER_SUCCESS: {
            const newState = {
                ...state,
                data: action.data,
                userRequest: false
            };

            if (action.accessToken) {
                newState.accessToken = action.accessToken;
            }

            return newState;
        }
        case GET_USER_FAILED: {
            return {
                ...state,
                userFailed: true,
                userRequest: false
            };
        }
        default: {
            return state;
        }
    }
};

export default UserReducer;
