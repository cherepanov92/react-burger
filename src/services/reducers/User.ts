import { deleteCookie } from '../../utils/helpers';

export const GET_USER_REQUEST = 'GET_USER_REQUEST';
export const GET_USER_SUCCESS = 'GET_USER_SUCCESS';
export const GET_USER_FAILED = 'GET_USER_FAILED';

export const LOGOUT_USER = 'LOGOUT_USER';

type userState = {
    data?: any;
    accessToken?: string;
};

export default function userReducer(state: userState = {}, action: any) {
    switch (action.type) {
        case GET_USER_REQUEST: {
            return {
                ...state,
                userRequest: true,
                userFailed: false
            };
        }
        case GET_USER_SUCCESS: {
            const newState = {
                ...state,
                data: action.data,
                userRequest: false,
                accessToken: state.accessToken
            };

            if (action.accessToken) {
                newState.accessToken = action.accessToken;
            }

            return newState;
        }
        case GET_USER_FAILED: {
            return {
                ...state,
                userFailed: true,
                userRequest: false
            };
        }
        case LOGOUT_USER: {
            deleteCookie('refreshToken');
            deleteCookie('accessToken');
            return { ...state, data: null, accessToken: null };
        }
        default: {
            return state;
        }
    }
};
