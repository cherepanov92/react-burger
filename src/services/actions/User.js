import { getUserApiData, login } from '../../utils/api';
import { APPEND_ERROR_MODAL_TYPE } from './Modal';
import { GET_USER_FAILED, GET_USER_REQUEST, GET_USER_SUCCESS } from '../reducers/User';

export const loginUser = (email, password) => {
    return function (dispatch) {
        dispatch({
            type: GET_USER_REQUEST
        });
        login(email, password)
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: GET_USER_SUCCESS,
                        name: res.user.name,
                        email: res.user.email
                    });
                } else {
                    dispatch({
                        type: GET_USER_FAILED
                    });
                }
            })
            .catch(err =>
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                })
            );
    };
};

export const getUserData = () => {
    return function (dispatch) {
        dispatch({
            type: GET_USER_REQUEST
        });
        getUserApiData()
            .then(res => {
                if (res && res.success) {
                    dispatch({
                        type: GET_USER_SUCCESS,
                        name: res.user.name,
                        email: res.user.email
                    });
                } else {
                    dispatch({
                        type: GET_USER_FAILED
                    });
                }
            })
            .catch(err =>
                dispatch({
                    type: APPEND_ERROR_MODAL_TYPE,
                    message: err.message
                })
            );
    };
};
