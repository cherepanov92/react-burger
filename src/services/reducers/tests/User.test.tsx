import React from 'react';
import renderer from 'react-test-renderer';

import { GET_USER_FAILED, GET_USER_REQUEST, GET_USER_SUCCESS, LOGOUT_USER } from '../../actions/User';
import { InitialUserState } from '../User';
import userReducer from '../User';
import { TUserInfoData } from '../../../utils/types';

const userItem: TUserInfoData = {
    email: 'test@test.te',
    name: 'test_boy'
};

const accessToken = 'iAmSuperAccessToken1337';

describe('UserReducer:', () => {
    it('Начальное состояние', () => {
        expect(userReducer(InitialUserState, {} as any)).toEqual(InitialUserState);
        const tree = renderer.create(<>Начальное состояние</>).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('Отправка запроса данных о пользователе', () => {
        expect(
            userReducer(InitialUserState, {
                type: GET_USER_REQUEST
            })
        ).toEqual({
            ...InitialUserState,
            userRequest: true
        });
    });

    describe('Получение данных о пользователе:', () => {
        it('С обновлением accessToken-а', () => {
            expect(
                userReducer(InitialUserState, {
                    type: GET_USER_SUCCESS,
                    data: userItem,
                    accessToken: accessToken
                })
            ).toEqual({
                ...InitialUserState,
                userRequest: false,
                data: userItem,
                accessToken: accessToken
            });
        });

        it('Без обновления accessToken-а', () => {
            expect(
                userReducer(InitialUserState, {
                    type: GET_USER_SUCCESS,
                    data: userItem,
                    accessToken: undefined
                })
            ).toEqual({
                ...InitialUserState,
                userRequest: false,
                data: userItem
            });
        });
    });

    it('Ошибка при запросе данных о пользователе', () => {
        expect(
            userReducer(InitialUserState, {
                type: GET_USER_FAILED
            })
        ).toEqual({
            ...InitialUserState,
            userFailed: true,
            userRequest: false
        });
    });

    it('Разлогирование пользователя', () => {
        expect(
            userReducer(InitialUserState, {
                type: LOGOUT_USER
            })
        ).toEqual({
            ...InitialUserState,
            data: null,
            accessToken: null
        });
    });
});
