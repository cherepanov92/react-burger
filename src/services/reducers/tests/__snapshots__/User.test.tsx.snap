// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserReducer: Начальное состояние 1`] = `"Начальное состояние"`;
